File: src/main/scala/Optimization1.scala
import scala.util.parsing.json._
import scala.io.Source
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.catalyst.expressions.{Literal, Multiply}
import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
import org.apache.spark.sql.catalyst.rules.Rule
/**
  * Created by Spark on 27/10/2016.
  */
object Optimization1 {

  def main(args: Array[String])  : Unit = {

    val sparkSession = SparkSession.builder.
      master("local")
      .appName("optimization1")
      .getOrCreate()

    val employees = sparkSession.read.option("header","true").csv("src/main/resources/employees.csv")
    val salaries = sparkSession.read.option("header","true").csv("src/main/resources/salaries.csv")
    val departments = sparkSession.read.option("header","true").csv("src/main/resources/departments.csv")
    val titles = sparkSession.read.option("header","true").csv("src/main/resources/titles.csv")
    val dept_emp = sparkSession.read.option("header","true").csv("src/main/resources/dept_emp.csv")
    val dept_manager = sparkSession.read.option("header","true").csv("src/main/resources/dept_manager.csv")

   /* // Parsing da file a formato map di string
    val jsonSource : String = Source.fromFile("src/main/resources/DB_config.json").getLines.mkString

    val result = JSON.parseFull(jsonSource)
    result match {
      // Matches if jsonStr is valid JSON and represents a Map of Strings to Any
      case Some(map: Map[String, Any]) => println(map)
      case None => println("Parsing failed")
      case other => println("Unknown data structure: " + other)
    }

    val path = "src/main/resources/salaries.csv"
    val db = sparkSession.read.option("header","true").csv(path)
    db.printSchema()*/

    /*val multipliedDF = employees.selectExpr("first_name")
      .join(salaries, employees ("emp_no") === salaries ("emp_no"))
      .where( employees ("gender") === "F")*/

    val conf = new Configurator
    conf.Builder()
    val nodi = conf.nodes
    println(nodi.get(1).getName)

    salaries.createOrReplaceTempView("salaries")
    employees.createOrReplaceTempView("employees")
    titles.createOrReplaceTempView("titles")
    val sqlDF = sparkSession.sql("SELECT first_name FROM salaries s Join employees e ON s.emp_no=e.emp_no WHERE gender=='F'")
    println(sqlDF.queryExecution.optimizedPlan.numberedTreeString)
    val sqlDF2 = sparkSession.sql("SELECT first_name FROM employees e JOIN (SELECT s.emp_no, salary FROM salaries s JOIN titles t " +
      "ON s.emp_no=t.emp_no GROUP BY s.salary HAVING t.title='Staff') st ON e.emp_no=st.emp_no GROUP BY first_name HAVING salary>70000")
    println(sqlDF2.queryExecution.optimizedPlan.numberedTreeString)

  }
}

File: src/main/java/Optimization1.java
/*
    ----------------------------------------------------------------------------------------------------
    "Spark Catalyst Customization for Distributed Systems"
    ----------------------------------------------------------------------------------------------------
    European Research Project “Enforceable Security in the Cloud to Uphold Data Ownership” (ESCUDO-CLOUD)
    ----------------------------------------------------------------------------------------------------
    Release 1.0 by Antonio Cosseddu and Giovanni B. Nava
    Release 2.0 by Giovanni B. Nava
    ----------------------------------------------------------------------------------------------------
 */

import AuthorizationModel.*;
import CostModel.*;
import DataConfigBuilder.DataBuilder;
import RelationProfileTreeBuilder.*;
import TreeStructure.BinaryNode;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import java.io.FileNotFoundException;


public class Optimization1
{

    public static void main(String[] args) throws FileNotFoundException
    {

        DataBuilder.getDataBuilder().buildData();

        // Create views for the query
        DataBuilder.getDataBuilder().salaries.createOrReplaceTempView("salaries");
        DataBuilder.getDataBuilder().employees.createOrReplaceTempView("employees");
        DataBuilder.getDataBuilder().titles.createOrReplaceTempView("titles");
        DataBuilder.getDataBuilder().departments.createOrReplaceTempView("departments");
        DataBuilder.getDataBuilder().dept_emp.createOrReplaceTempView("dept_emp");
        DataBuilder.getDataBuilder().dept_manager.createOrReplaceTempView("dept_manager");

        // QUERY
        Dataset<Row> sqlDF = DataBuilder.getDataBuilder().sparkSession.sql("SELECT first_name FROM salaries s Join employees e ON s.emp_no=e.emp_no WHERE salary > 70000 GROUP BY first_name");

        // Generate the relation tree
        RelationProfileTree tree = new RelationProfileTree(sqlDF.queryExecution().optimizedPlan());

        // Generate the tree for the authorized subjects
        AuthorizationModel model = new AuthorizationModel(DataBuilder.getDataBuilder().providers, tree.getRelationTree());

        // produce l'albero ottimizzato numerato
        //System.out.println(sqlDF.queryExecution().optimizedPlan().numberedTreeString());

        // istruzioni per stampare gli operatori di ogni operazione
        //System.out.println(sqlDF.queryExecution().optimizedPlan().apply(4).expressions().toList().apply(0));
        //System.out.println(sqlDF.queryExecution().optimizedPlan().apply(2).constraints().toList());
        //System.out.println(sqlDF.queryExecution().optimizedPlan().apply(4).constraints().toList().apply(1).prettyName());
        //System.out.println(sqlDF.queryExecution().optimizedPlan().apply(4).constraints().toList().apply(1).flatArguments().toList().apply(1));

        // COST TEST of JOIN
        BinaryNode<Relation> newNode = tree.getRelationTree().getRoot().getLeft().getLeft().getLeft().getLeft();

        CostModel costModel = new CostModel(DataBuilder.getDataBuilder().providers, tree);
//        double cost = costModel.computeCost(DataBuilder.getDataBuilder().providers.get(0), DataBuilder.getDataBuilder().providers.get(1), newNode);
//        System.out.println("---> Da " + DataBuilder.getDataBuilder().providers.get(0).getName() + " a " + DataBuilder.getDataBuilder().providers.get(1).getName());
//        System.out.println("---> OPERATION = " + newNode.getElement().getOperation());
//        System.out.println("---> COSTO = " + cost);
//
//        cost = costModel.computeCost(DataBuilder.getDataBuilder().providers.get(1), DataBuilder.getDataBuilder().providers.get(2), newNode);
//        System.out.println("---> Da " + DataBuilder.getDataBuilder().providers.get(1).getName() + " a " + DataBuilder.getDataBuilder().providers.get(2).getName());
//        System.out.println("---> OPERATION = " + newNode.getElement().getOperation());
//        System.out.println("---> COSTO = " + cost);
//
//        cost = costModel.computeCost(DataBuilder.getDataBuilder().providers.get(2), DataBuilder.getDataBuilder().providers.get(0), newNode);
//        System.out.println("---> Da " + DataBuilder.getDataBuilder().providers.get(2).getName() + " a " + DataBuilder.getDataBuilder().providers.get(0).getName());
//        System.out.println("---> OPERATION = " + newNode.getElement().getOperation());
//        System.out.println("---> COSTO = " + cost);


        // Create a new EncryptionProfile
        EncryptionProfile encProfile = new EncryptionProfile();
        // Init the EncryptionProfile with all default encryptions
        encProfile.setupDefault(tree.getRelationTree());

        PlansMap plansMap = costModel.generatePlans(tree.getRelationTree().getRoot(), encProfile);

        Plan optimalPlan = costModel.getOptimalPlan(plansMap);

        System.out.println("*****************************************************");
        System.out.println("> [INFO] OPTIMAL PLAN:");
        System.out.println(optimalPlan.toString());
        System.out.println("*****************************************************");

    }

}